### PR TITLE
non-traitors can no longer insert tc into uplinks

### LINKED ISSUE #14338: You can 'spycheck' tators by inserting TC into their uplinks
Originally reported by yglop#5593

Reproduce:
1. Get 5 TC from captain fox (or from any other way)
2. Try insert one TC into someone PDA. If it would insert - player is valid

### PR DESCRIPTION
**Pull request by deltanedas**
## About the PR
you can no longer valid check pdas by inserting tc from renault, only traitors can.
this fixes #14338

doesn't affect nukie uplinks

**Media**
- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**
:cl:
- tweak: Only traitors can insert TC into PDAs now.

### CODE REVIEW

**Comment by 0x6273** [comment]
> also, nukie uplinks cannot be loaded with tc (though still used). not sure about how to fix this, or if it should be fixed

Can't you just check if AccountOwner on StoreComponent is null? It's null for nukies and spawned uplinks, but not traitors.

**Comment by deltanedas** [comment]
ill check that out

**Comment by Cheackraze** [comment]
As long as nukies can still add tc, they share tc with eachother more often than traitors do

**Comment by deltanedas** [comment]
this pr isnt meant to affect nukies at all, its just to prevent the valid check

File: Content.Server/Store/Components/StoreComponent.cs [hunk 1]
```diff
@@ -66,6 +66,12 @@ public sealed class StoreComponent : Component
     [DataField("buySuccessSound")]
     public SoundSpecifier BuySuccessSound = new SoundPathSpecifier("/Audio/Effects/kaching.ogg");
     #endregion
+
+    /// <summary>
+    /// Is this store a hidden traitor uplink (not nukie), if it is then only traitors can insert currency to prevent valid checking pdas with renault's help.
+    /// </summary>
+    [DataField("traitor")]
+    public bool Traitor;
```
Comment (AJCM-git) [inline Content.Server/Store/Components/StoreComponent.cs:74]: the store shouldnt have a notion of traitors, make this a whitelist of roles or something

**Review by AJCM-git — CHANGES_REQUESTED** [review]

File: Content.Shared/Store/StorePresetPrototype.cs [hunk 2]
```diff
@@ -38,4 +38,10 @@ public sealed class StorePresetPrototype : IPrototype
     /// </summary>
     [DataField("currencyWhitelist", customTypeSerializer: typeof(PrototypeIdHashSetSerializer<CurrencyPrototype>))]
     public HashSet<string> CurrencyWhitelist { get; } = new();
+
+    /// <summary>
+    /// Is this store a hidden traitor uplink (not nukie), if it is then only traitors can insert currency to prevent valid checking pdas with renault's help.
+    /// </summary>
+    [DataField("traitor")]
+    public bool Traitor;
```
Comment (AJCM-git) [inline Content.Shared/Store/StorePresetPrototype.cs:46]: same

File: Content.Server/Store/Components/StoreComponent.cs (hunk 1 above)
Comment (AJCM-git) [reply Content.Server/Store/Components/StoreComponent.cs:74]: Nevermind, i didnt realize HasRole was VERY old, make this an AntagWhitelist field or something (name due to change) and check if AntagPrototype IDs match

Comment (deltanedas) [reply Content.Server/Store/Components/StoreComponent.cs:75]: dont think todos go in summary so put on its own comment

**Comment by mirrorcult** [comment]
now you can valid check people by asking them to insert TC into a PDA?

**Comment by mirrorcult** [comment]
a better solution imo is to just only allow inserting TC if the uplink is open

**Comment by Interrobang01** [comment]
> now you can valid check people by asking them to insert TC into a PDA?

wouldn't the easy counter to that strategy be to just... not click the pda? I think it's fine, as long as we don't implement a way for players to see what other players click.

**Comment by deltanedas** [comment]
> a better solution imo is to just only allow inserting TC if the uplink is open

how would i check if uplink is open when insert attempt is raised

**Comment by deltanedas** [comment]
for some reason `SessionHasOpenUi` always returns false... lovely

File: Content.Server/Store/Systems/StoreSystem.cs [hunk 3]
```diff
@@ -67,6 +69,12 @@ private void OnAfterInteract(EntityUid uid, CurrencyComponent component, AfterIn
         if (args.Target == null || !TryComp<StoreComponent>(args.Target, out var store))
             return;
 
+        // require the store to be open before inserting currency
+        var user = args.User;
+        Logger.DebugS("TEMP_error", $"open: {_ui.SessionHasOpenUi(uid, StoreUiKey.Key, Comp<ActorComponent>(user).PlayerSession)}");
+        if (!TryComp<ActorComponent>(user, out var actor) || !_ui.SessionHasOpenUi(uid, StoreUiKey.Key, actor.PlayerSession))
+            return;
```
Comment (AJCM-git) [inline Content.Server/Store/Systems/StoreSystem.cs:76]: Now that uplinks depend on ringtone you can just check for that instead right? (also check if not-pda uplinks still work)

**Review by AJCM-git — CHANGES_REQUESTED** [review]

File: Content.Server/Store/Systems/StoreSystem.cs (hunk 3 above)
Comment (deltanedas) [reply Content.Server/Store/Systems/StoreSystem.cs:76]: i think that would instead just be integrated with uplink locking/unlocked if/when its done, makes mores sense than explicitly checking ringtone here

**Pull request merged**